<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/images/IDSN-logo.png" width="200">

# ETL Using Apache Spark

**Estimated time needed:** 20 minutes

The purpose of this lab is to show you how to pull data from a source system, transform it and store it back to the target system.

## Objectives

After completing this lab, you will be able to:

1.  Pull data from the HMP dataset.
2.  Create a Spark Dataframe from the raw data.
3.  Write the Dataframe to CSV.
4.  Convert the CSV data to PARQUET.
5.  Condense PARQUET to a single file.
6.  Upload the Parquet file to Cloud Object Store

But don’t be scared, we provide you with a set of sample notebooks you can modify and hook together. The library where you can get the notebooks from is called CLAIMED – the Component Library for AI, Machine Learning, ETL, and Data Science and is an open-source library available on the [CLAIMED repo](https://github.com/IBM/claimed/tree/master/component-library).

You’ll use [Elyra – a JupyterLab extension](https://elyra.readthedocs.io/en/stable/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01) for editing the notebooks; and, if you like you can use the [pipeline editor](https://elyra.readthedocs.io/en/latest/getting_started/overview.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01#ai-pipelines-visual-editor) of Elyra to visually join them into a data pipeline.

Elyra is the foundation of the [IBM Watson Studio Pipelines](https://medium.com/ibm-data-ai/automating-the-ai-lifecycle-with-ibm-watson-studio-orchestration-flow-4450f1d725d6?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01) tool which can be used in the cloud. Feel free to give it a shot as well to experience how a business user would do the job.

We’ll use the HMP dataset. The dataset is publically available [here](https://github.com/wchill/HMP_Dataset).

In previous generation BigData systems, HDFS was the core data store. Nowadays, S3 compatible Cloud Object Storage (COS) is the de-facto standard across clouds and also starts to get traction in local data centers (Ceph, Minio).

So let’s get started with the lab!

## Exercise 1 : Import the CLAIMED library to JupyterLab

1.  In a separate browser tab, please open the CLAIMED component library: [GitHub Link](https://github.com/IBM/claimed/tree/master/component-library)
2.  Don’t hesitate to give us a star (1) :), then please click the Fork button. (2)

(Taking this action enables you to work using your own copy - you then can replace the link to your fork below in the next cell)

![Elyra Github repo](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/fp4.png)

Now let's clone (download) the library:


In [1]:
%%bash
rm -Rf claimed
git clone https://github.com/IBM/claimed.git
cd claimed

Cloning into 'claimed'...
Checking out files: 100% (434/434), done.


Congratulations, you’ve successfully imported the component library.

## Exercise 2 : Explore the CLAIMED component library

Now it’s time to familiarize yourself a bit with some components (Jupyter Notebooks) in the component library.

1.  In JuypterLab, please go to the file explorer (1), double-click on folder "claimed" and then double click on the folder “component-library” (2)

![jupyerlab exploration](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/fp7.png)

2.  Please open the folder called “transform” (1) and open the notebook called spark-csv-to-parquet (2). The prefix “spark” indicates that the notebook is using Apache Spark to perform its task. From the name you can deduce that this component transforms a file from “parquet” to “csv” format. Each notebook starts with a title and description of what it is supposed to do, followed by commands to install library dependencies. Then, a set of parameters this notebook accepts is provided (3), followed by an actual implementation of pulling those parameters from the environment (4). Finally, the actual task is implemented in source code (5).

![jupyerlab exploration](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/fp8.png)

3.  Please explore the other components in the library and look at how they are implemented. They serve as cookie-cutters, or examples, for an abundance of daily data science tasks and hopefully you can learn from them.


## Exercise 3 : Start the ETL process

### Task A : Pull the data


Now we need to pull data from a remote github repository and convert it to CSV.
You can open the notebook called *input-hmp.ipynb* in folder *claimed/component-library/input* and run
each cell top down, one by one. Alternatively, the following cell is doing the job as well.
If you run the code by executing the cell below, please open the notebook in parallel and
investigate how it was implemented.

Please note, that that way we can pass a parameter called *sample* which reduces data size and processing
time. Sampling is often done on "Non Big Data" Systems, but in Apache Spark this is not necessary most of the
times because Apache Spark can handle any amount of data - so sampling is done here only for saving
time during the lab.

Note: A cell which creates a lot of output can be switched into scrolling mode. Just right-click into the output canvas (1) and click on “Enable Scrolling for Outputs” (2)

![creating a pipeline](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/fp10.png)

3.  After quite some time you should see a folder called “data.csv” in the “data” directory you’ve previously created. Please note that Apache Spark always creates folders containing individual files, one per partition. This is not a problem because Spark doesn’t distinguish between files and folders and threats folders as they were files. (The only way to get a file is to repartition the data frame to one and extract/rename the file inside the folder df = df.repartition(1))


In [1]:
!ipython ./claimed/component-library/input/input-hmp.ipynb data_dir=./data/ sample=0.01

]0;IPython: labs/BD0231EN37
Starting installation...
Please <<RESTART YOUR KERNEL>> (Kernel->Restart Kernel and Clear All Outputs)
23/03/26 12:47:54 WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
23/03/26 12:47:56 WARN util.Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
Cloning into 'HMP_Dataset'...
remote: Enumerating objects: 865, done.
remote: Total 865 (delta 0), reused 0 (delta 0), pack-reused 865
Receiving objects: 100% (865/865), 1010.96 KiB | 583.00 KiB/s, done.
Checking out files: 100% (848/848), done.
Skipping: Accelerometer-2011-04-11-13-28-18-brush_teeth-f1.txt
Skipping: Accelerometer-2011-04-11-13-29-54-brush_teeth-f1.txt
Skipping: Accelerometer-2011-05-30-08-35-11-brush_teeth-f1.txt
Skipping: Accelerometer-2011-05-30-09-36-50-brush_

### Task B : Convert from CSV to PARQUET

CSV isn't a memory and storage efficient file format - the same holds for JSON by the way. Therefore,
formats like [Avro](https://en.wikipedia.org/wiki/Apache_Avro?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01), [Parquet](https://en.wikipedia.org/wiki/Apache_Parquet?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01), [ORC](https://en.wikipedia.org/wiki/Apache_ORC?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01) or [RCFile](https://en.wikipedia.org/wiki/RCFile?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01) emerged. Therefore we convert the CSV data to PARQUET before uploading it to Cloud Object Store.
Again, [CLAIMED](https://github.com/IBM/claimed/tree/master/component-library) provides a component as jupyter notebook
exactly doing that. Please explore the notebook under *component-library/transform/spark-csv-to-parquet.ipynb*. You can either directly run the notebook or execute the following cell.


In [2]:
!ipython ./claimed/component-library/transform/spark-csv-to-parquet.ipynb data_dir=./data/

]0;IPython: labs/BD0231EN37
Starting installation...
Please <<RESTART YOUR KERNEL>> (Kernel->Restart Kernel and Clear All Outputs)


### Task C: Condense parquet file

As mentioned before - Apache Spark creates folders instead of files - in the folder, besides some status information, you will see that one sub-file per partition is written. This is handy for parallel processing but sometimes we
want all data to be self-contained in a single file. Therefore, the "spark-condense-parquet.ipynb" component
in the *./claimed/component-library/transform/* folder does this job. Let's execute it using the cell below, but please
also open the notebook to see what it is doing - alternatively you can also run the notebook cell by cell.


In [1]:
!ipython ./claimed/component-library/transform/spark-condense-parquet.ipynb data_dir=./data/

]0;IPython: labs/BD0231EN37
Starting installation...
Please <<RESTART YOUR KERNEL>> (Kernel->Restart Kernel and Clear All Outputs)
23/03/26 12:53:40 WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
23/03/26 12:53:42 WARN util.Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
---------------------------------------------------------------------------
Py4JJavaError                             Traceback (most recent call last)
~/conda/envs/python/lib/python3.7/site-packages/pyspark/sql/utils.py in deco(*a, **kw)
     62         try:
---> 63             return f(*a, **kw)
     64         except py4j.protocol.Py4JJavaError as e:

~/conda/envs/python/lib/python3.7/site-packages/py4j/protocol.py in get_return_value(answer, gateway_client, target_id, name)
  

### Task D : Obtain access to Cloud Object Store

1.  Please create an account on [IBM Cloud](https://cloud.ibm.com/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01) - it is free, doesn't expire and no credit card is required. You also get 25 GB storage on Cloud Object Store (COS) for free. IBM COS is compatible to S3. So what you learn here will work on all S3 compliant storage system in different clouds and also in your local data center.

2.  Once you create and verify your account you can create an S3 compliant Object Storage service by clicking the following [link](https://cloud.ibm.com/objectstorage/create?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01) - please select the lite plan as it includes 25 GB of free storage

![Create COS](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/ibm_cos_create.png)

3.  You can always go back to the service created by accessing the [IBM Cloud Service Resource List](https://cloud.ibm.com/resources?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01). Just look for an entry in the table under the "Storage" category and click on it.

![Resource List](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/ibm_resource_list_cos.png)

4.  Now please click "Create Bucket"

![Create Bucket](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/ibm_cos_create_bucket.png)

5.  Select the "Quickly get started" option by clicking the arrow

![Create Bucket_2](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/ibm_cos_create_bucket_2.png)

6.  Click on "next" twice, then click "View Bucket Configuration"

7.  Here, please note the bucket name, and the location, then click "Service Credentials"

![Create SC](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/ibm_cos_location_sc.png)

8.  Click on "New Credential", then under "Advanced Options", please ensure that "Include HMAC Credential" is active. Then click "Add"

![hmac](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/hmac.png)

9.  Please open the newly created credentials, an in the "cos_hmac_keys" section, please note down "access_key_id" and "secret_access_key"

![sc](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/sc.png)

10. Click on "Endpoints", then select the appropriate location you've obtained four steps before. That way the correct public endpoint is displayed, please note it somewhere for future use.

![endpoint](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/endpoint.png)

Congratulations! You've obtained all information to store your date to Cloud Object Storage


### Task E: Upload file to Cloud Object Storage

Now it's time to use the data you've gathered from task B and fill in the following variables in the next cell.
Then please run the cell. This will upload the parquet file you've created to Cloud Object Store where is it
stored at the lowest possible cost and made available to others.

<span style="color:red">Please note: The endpoint you need to set is not the "endpoints" from the service credentials section but from Task D Step 11</span>


> Note :After executing the below code, if you  get an error as:

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. 

This behaviour is the source of the following dependency conflicts.aiobotocore 1.4.1 requires botocore<1.20.107,>=1.20.106,
but you have botocore 1.10.62 which is incompatible.
```

Kindly ignore the error and proceed to **TaskF** .Your file will be deployed.


In [1]:
%%bash
export access_key_id='d345c4c5d0174355b2b8e1eb46ad6b2a'
export secret_access_key='7443264455a50aef3fad9770503d2f2ae0343e06f1d432dc'
export endpoint='https://control.cloud-object-storage.cloud.ibm.com/v2/endpoints'
export bucket_name='cloud-object-storage-cos-standard-2ns'
export source_file='data_condensed.parquet'
export destination_file='data.parquet'
export data_dir='./data/'
ipython ./claimed/component-library/output/upload-to-cos.ipynb $access_key_id $secret_access_key $endpoint $bucket_name $source_file $destination_file $data_dir

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.7/72.7 kB 8.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 948.1/948.1 kB 53.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.4/10.4 MB 93.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.0/143.0 kB 14.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 66.8/66.8 kB 10.4 MB/s eta 0:00:00
INFO: pip is looking at multiple versions of fsspec to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of s3fs to determine which version is compatible with other requirements. This could take a while.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 66.8/66.8 kB 8.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.8/65.8 kB 9.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.5/139.5 kB 21.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 138.8/138.8 kB 20.1 MB/s et

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
boto3 1.7.62 requires botocore<1.11.0,>=1.10.62, but you have botocore 1.29.76 which is incompatible.


CalledProcessError: Command 'b"export access_key_id='d345c4c5d0174355b2b8e1eb46ad6b2a'\nexport secret_access_key='7443264455a50aef3fad9770503d2f2ae0343e06f1d432dc'\nexport endpoint='https://control.cloud-object-storage.cloud.ibm.com/v2/endpoints'\nexport bucket_name='cloud-object-storage-cos-standard-2ns'\nexport source_file='data_condensed.parquet'\nexport destination_file='data.parquet'\nexport data_dir='./data/'\nipython ./claimed/component-library/output/upload-to-cos.ipynb $access_key_id $secret_access_key $endpoint $bucket_name $source_file $destination_file $data_dir\n"' returned non-zero exit status 1.

### Task F: Double-check if the file arrived at its destination

To make sure everything worked as designed (actually an unnecessary step if you haven't received any errors)
we open the IBM Cloud user interface. Please open the bucket contents as you've learned in Task C and check if
the file arrived as shown below.

![cos_contents](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-BD0231EN-SkillsNetwork/labs/images/check_cos.png)

Congratulations, this concludes this lab.


### Task F: (Optional) Pull another file and upload it as well to COS

To make things easier let's first flush the data dir:


In [3]:
!rm -Rf ./claimed/data/*

Now it's time for you to pull some other data set. You can choose anything you like but you can also use
the "input-pardata.ipynb" notebook in folder "./component-library/input". This component pulls the JFK
weather data set as data.csv. From there, please convert this CSV file to PARQUET, condense it and upload
it to Object Storage as "jfk_data.parquet". You can either directly open the notebooks, run it inline this
notebook below or even create a pipeline using the [Elyra Pipeline editor](https://youtu.be/0SWRMnhIZMA?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01&t=1810).

We have provided an inline solution for you below in case you get stuck.


In [4]:
!ipython ./claimed/component-library/input/input-pardata.ipynb data_dir=./data/

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.9/43.9 kB 4.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 45.9 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 50.5 MB/s eta 0:00:00a 0:00:01
  Attempting uninstall: Pillow
    Found existing installation: Pillow 8.1.0
    Uninstalling Pillow-8.1.0:
      Successfully uninstalled Pillow-8.1.0
]0;IPython: labs/BD0231EN---------------------------------------------------------------------------
IsADirectoryError                         Traceback (most recent call last)
/resources/labs/BD0231EN/claimed/component-library/input/input-pardata.ipynb in <module>
----> 1 df.to_csv(data_dir + data_csv, index=False)

~/conda/envs/python/lib/python3.7/site-packages/pandas/core/generic.py in to_csv(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, line_terminator, chunksize, date_format, doublequote, esca

In [5]:
!ipython ./claimed/component-library/transform/spark-csv-to-parquet.ipynb data_dir=./data/

]0;IPython: labs/BD0231EN37
Starting installation...
Please <<RESTART YOUR KERNEL>> (Kernel->Restart Kernel and Clear All Outputs)


In [ ]:
!ipython ./claimed/component-library/transform/spark-condense-parquet.ipynb data_dir=./data/

]0;IPython: labs/BD0231EN37
Starting installation...
Please <<RESTART YOUR KERNEL>> (Kernel->Restart Kernel and Clear All Outputs)
23/03/25 23:07:41 WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [1]:
%%bash
export access_key_id='d345c4c5d0174355b2b8e1eb46ad6b2a'
export secret_access_key='7443264455a50aef3fad9770503d2f2ae0343e06f1d432dc'
export endpoint='https://control.cloud-object-storage.cloud.ibm.com/v2/endpoints'
export bucket_name='cloud-object-storage-cos-standard-2ns'
export source_file='data_condensed.parquet'
export destination_file='jfk_data.parquet'
export data_dir='./data/'
ipython ./claimed/component-library/output/upload-to-cos.ipynb $access_key_id $secret_access_key $endpoint $bucket_name $source_file $destination_file $data_dir

]0;IPython: labs/BD0231ENRequirement already satisfied: aiobotocore in /home/jupyterlab/conda/envs/python/lib/python3.7/site-packages (2.5.0)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/resources/labs/BD0231EN/claimed/component-library/output/upload-to-cos.ipynb in <module>
----> 1 s3.put(data_dir + source_file, bucket_name + '/' + destination_file)

~/conda/envs/python/lib/python3.7/site-packages/fsspec/asyn.py in wrapper(*args, **kwargs)
     86     def wrapper(*args, **kwargs):
     87         self = obj or args[0]
---> 88         return sync(self.loop, func, *args, **kwargs)
     89 
     90     return wrapper

~/conda/envs/python/lib/python3.7/site-packages/fsspec/asyn.py in sync(loop, func, timeout, *args, **kwargs)
     67         raise FSTimeoutError
     68     if isinstance(result[0], BaseException):
---> 69         raise result[0]
     70     return result[0]
     7

CalledProcessError: Command 'b"export access_key_id='d345c4c5d0174355b2b8e1eb46ad6b2a'\nexport secret_access_key='7443264455a50aef3fad9770503d2f2ae0343e06f1d432dc'\nexport endpoint='https://control.cloud-object-storage.cloud.ibm.com/v2/endpoints'\nexport bucket_name='cloud-object-storage-cos-standard-2ns'\nexport source_file='data_condensed.parquet'\nexport destination_file='jfk_data.parquet'\nexport data_dir='./data/'\nipython ./claimed/component-library/output/upload-to-cos.ipynb $access_key_id $secret_access_key $endpoint $bucket_name $source_file $destination_file $data_dir\n"' returned non-zero exit status 1.

## Author(s)

[Romeo Kienzler](https://ibm.com/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01)

[Karthik Muthuraman](https://ibm.com/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMBD0231ENSkillsNetwork26766988-2022-01-01)

## Changelog

| Date       | Version | Changed by         | Change Description       |
| ---------- | ------- | ------------------ | ------------------------ |
| 2021-05-12 | 0.1     | Romeo Kienzler     | Initial version created  |
| 2021-06-22 | 1.0     | Karthik Muthuraman | First editorial pass     |
| 2021-05-12 | 1.1     | Romeo Kienzler     | Adjusted for API changes |
| 2021-05-26 | 1.2     | Romeo Kienzler     | Moved to jupyter lab     |
| 2021-05-26 | 1.3     | Romeo Kienzler     | Add optional task        |
| 2021-12-06 | 1.4     | Lakshmi Holla      | Added the warning note   |
|            |         |                    |                          |

## <h3 align="center"> © IBM Corporation 2021. All rights reserved. <h3/>
